install/terraform/modules/gke/variables.tf

```diff
@@ -25,11 +25,12 @@ variable "cluster" {
   type        = map
 
   default = {
-    "zone"             = "us-west1-c"
-    "name"             = "test-cluster"
-    "machineType"      = "n1-standard-4"
-    "initialNodeCount" = "4"
-    "project"          = "agones"
-    "network"          = "default"
+    "zone"              = "us-west1-c"
+    "name"              = "test-cluster"
+    "machineType"       = "n1-standard-4"
+    "initialNodeCount"  = "4"
+    "project"           = "agones"
+    "network"           = "default"
+    "kubernetesVersion" = "1.15"
```

Comment:
Question: Since this is a map, doesn't that mean users have to set all these values, they can't choose individual ones?

In which case the default value isn't actually useful.

It may be worth doing a slow transition away from a map here? @chrisst I think you had thoughts here.

Reply:
The issue is that maps get blindly replaced so if you don't provide `network` but override other variables then the line `var.cluster["network"]` will fail as there is no network supplied.

The simplest change is to flatten the object to individual fields but that would be a breaking change for consumers. You can convert to using locals internally and set the defaults in those based on the content of the `cluster` variable but it will require a bit of magic internally. https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/ has some good examples of setting locals based on input variables and merging in defaults.

Reply:
@chrisst Thank you for an explanation on how we can fix this replacement in maps.

Reply:
@chrisst Applied your comment, used lookup with locals.

Reply:
@chrisst - mind taking a look, make sure this is lining up with your proposed solution?

Reply:
Looks much more flexible! The `defaults` section also now just serves mostly as documentation unless a user completely omits the variable. Since the locals section serves as the source of truth in a partially defined block you'll want to keep these two in sync.

Reply:
Awesome! This makes me happy.